Create picocli extension (#9180)
quarkusio/quarkus: docs/src/main/asciidoc/picocli.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Command Mode with Picocli
:extension-status: experimental

include::./attributes.adoc[]

https://picocli.info/[Picocli] is an open source tool for creating rich command line applications.

Quarkus provides support for using Picocli. This guide contains examples of `picocli` extension usage.

include::./status-include.adoc[]

> IMPORTANT: If you are not familiar with the Quarkus Command Mode, consider reading the link:command-mode-reference[Command Mode Applications] first.

Reviewer: ```suggestion
IMPORTANT: If you are not familiar with the Quarkus Command Mode, consider reading the link:command-mode-reference[Command Mode reference guide] first.
```

== Configuration

Once you have your Quarkus project configured you can add the `picocli` extension
to your project by running the following command in your project base directory.

[source,bash]
----
./mvnw quarkus:add-extension -Dextensions="picocli"
----

This will add the following to your pom.xml:

[source,xml]
----
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-picocli</artifactId>
    </dependency>
----

== Simple Picocli Application with one Command

> Simple PicocliApplication with only one Command can be created as follows:

Reviewer 2: ```suggestion
Simple Picocli Application with only one Command can be created as follows:
```

Reviewer 2: not really fan of calling this "Picocli Application" it is still an Quarkus Application just enhanced with command line parsing done by picocli.

Reviewer 3: How about something like this:

> == Standalone picocli-based command line application
> A standalone command line application with a single picocli command can be created as follows:

I'm finding the term "standalone" command/application, useful when distinguishing from a "command suite" with subcommands.

Reviewer 4: > not really fan of calling this "Picocli Application" it is still an Quarkus Application just enhanced with command line parsing done by picocli.

+1

> How about something like this:
> 
> > == Standalone picocli-based command line application
> > A standalone command line application with a single picocli command can be created as follows:
> 
> I'm finding the term "standalone" command/application, useful when distinguishing from a "command suite" with subcommands.

But can't we also create subcommands with this extension? I think calling simply a `command line application` may be good enough

Reviewer 3: > But can't we also create subcommands with this extension?

I was thinking to have a separate section in the documentation for a command with subcommands; both are common scenarios and it is helpful to have clear examples for both.

[source,java]
----
package com.acme.picocli;

import picocli.CommandLine;

import javax.enterprise.context.Dependent;
import javax.inject.Inject;

@CommandLine.Command // <1>
public class HelloCommand implements Runnable {

    @CommandLine.Option(names = {"-n", "--name"}, description = "Who will we greet?", defaultValue = "World")
    String name;

    private final GreetingService greetingService;

    @Inject // <2>
    public HelloCommand(GreetingService greetingService) {
        this.greetingService = greetingService;
    }

    @Override
    public void run() {
        greetingService.sayHello(name);
    }
}

@Dependent
class GreetingService {
    void sayHello(String name) {
        System.out.println("Hello " + name + "!");
    }
}
----
<1> If there is only one class annotated with `picocli.CommandLine.Command` it will be used as entry point to Picocli CommandLine.

> <2> All classes annotated with `picocli.CommandLine.Command` are register as beans in CDI.

Reviewer: ```suggestion
<2> All classes annotated with `picocli.CommandLine.Command` are registered as CDI beans.
```

== Picocli Application with multiple Commands

> When multiple classes has `picocli.CommandLine.Command` annotation, then one of them need to be also annotated with `io.quarkus.picocli.runtime.annotations.TopCommand`.

Reviewer: ```suggestion
When multiple classes have the `picocli.CommandLine.Command` annotation, then one of them needs to be also annotated with `io.quarkus.picocli.runtime.annotations.TopCommand`.
```

> When multiple classes has `picocli.CommandLine.Command` annotation, then one of them need to be also annotated with `io.quarkus.picocli.runtime.annotations.TopCommand`.

Reviewer 2: having `@TopCommand` does not feel right - could it not reuse @QuarkusMain logic or at least similar logic to pick the main one ?

Reviewer 2: maybe just use @Named ?

Reviewer 3: I also have mixed feelings about `@TopCommand` (but perhaps I just need to get used to it...)
My preference would be that applications would always get access to the `CommandLine` instance, so they can configure it as needed and explicitly call `execute` on it (or `parseArgs` if they want to).
My approach was to skip `PicocliCommandLineFactory` altogether and let application code look like this:

```java
@QuarkusMain
@ApplicationScoped
@Command(name = "demo", mixinStandardHelpOptions = true)
public class ExampleApp implements Runnable, QuarkusApplication {
    @Inject
    PicocliBeansFactory factory;

    @Override
    public void run() {
        // business logic
    }

    @Override
    public int run(String... args) throws Exception {
        System.setProperty("quarkus.banner.enabled", "false");
        return new CommandLine(this, factory).execute(args);
    }
}
```

Author: @remkop Your example will work with current implementation. If user provide own `QuarkusMain` annotated class then `picocli` extension will not add `PicocliRunner` bean. I will add this to documentation.

I wanted simplest Quarkus/Picocli application to look like basic REST example. One, `Command` annotated, class could be good start for simple command line application.

As for `TopCommand`:
- I don't think that I can reuse `QuarkusMain` here, without big changes in `MainClassBuildStep`. `Command` class may just be `Object` and don't need to implement `QuarkusApplication`
- @maxandersen Using `Named` doesn't fell right also, since any `Object` can be annotated with `Command` and `Command` itself is not `Qualifier`.
- We could add `name` to `TopCommand` and have similar logic as in `QuarkusMain`. In fact there is already `quarkus.picocli.top-command` property which will work in similar way. It takes FQCN of class which will be used as entry point. Not sure if this property should be runtime or build time configuration. For now it is runtime.
- We could drop `TopCommand` all together and add `PicocliRunner` if there is only one `Command` annotated class. In all other cases user would be forced to provided `QuarkusMain` on its own.
- Or just drop `PicocliRunner` entirely as @remkop suggested.

Reviewer 5: Personally I like the annotation based approach. I think you can make the @QuarkusMain integration work without major MainClassBuildStep changes by looking for classes that have both @QuarkusMain and @Command. If one is found then make PicoliRunner use that class, and create a QuarkusApplicationClassBuildItem so that Quarkus won't look for the QuarkusMain class itself.

Reviewer 3: @stuartwdouglas, if I understand correctly, you propose that instead of introducing a new `@TopCommand` annotation, applications would use both the combination of the `@QuarkusMain` and the `@Command` annotation to specify a main class. That makes sense to me.

Reviewer 5: Yea. 

The only real downside to this is that it could create confusion, as it is a 3rd possible way to use @QuarkusMain, but it is still doing pretty much the same thing so I think it should be ok.

Author: @stuartwdouglas So picocli extension would need to have copy of logic for finding `@QuarkusMain` and if this Main is also annotated with `Command` then it would produce QuarkusApplicationClassBuildItem pointing to `PicocliRunner`. But, I think that with such approach `PicocliRunner` would need to be generated class. Both `@QuarkusMain` and `@Command` are not Qualifiers, so I'm not sure how to pass it to current `PicocliRunner`. This would also mean that QuarkusApplicationClassBuildItem has priority. If it exist then MainClassBuildStep will always use it. 

Note that current approach, with `@TopCommand` qualifiers, also allows things like this (after I add `METHOD` `@Target`): 

```java
@ApplicationScoped
public class Config {

    @Produces
    @TopCommand
    @IfBuildProfile("dev")
    public Object devCommand() {
        // Since this is Object, it can be also Class<?>
        // Instance will be created by CommandLine.IFactory
        return DevCommand.class;
    }

    @Produces
    @TopCommand
    @IfBuildProfile("prod")
    public Object prodCommand() {
        return ProdCommand.class;
    }

}
````

Reviewer 5: That is a good point that this allows for producer methods. It also allows for the 'Right click and run from IDE' use case, if you do Quarkus.run(PicocliRunner.class) in your main method, which my approach would not allow (we should probably document this). 

Thinking about it I think that the current approach is probably best.

> This can be overwritten with `quarkus.picocli.top-command` property.

Reviewer: ```suggestion
This can be overwritten with the `quarkus.picocli.top-command` property.
```

[source,java]
----
package com.acme.picocli;

import io.quarkus.picocli.runtime.annotations.TopCommand;
import picocli.CommandLine;

@TopCommand
@CommandLine.Command(mixinStandardHelpOptions = true, subcommands = {HelloCommand.class, GoodByeCommand.class})
public class EntryCommand {
}

@CommandLine.Command(name = "hello", description = "Greet World!")
class HelloCommand implements Runnable {

    @Override
    public void run() {
        System.out.println("Hello World!");
    }
}

@CommandLine.Command(name = "goodbye", description = "Say goodbye to World!")
class GoodByeCommand implements Runnable {

    @Override
    public void run() {
        System.out.println("Goodbye World!");
    }
}
----

== Customizing Picocli CommandLine instance

> You can customize CommandLine classes used by `picocli` extension by producing your own bean instance:

Reviewer: ```suggestion
You can customize CommandLine classes used by the `picocli` extension by producing your own bean instance:
```

[source,java]
----
package com.acme.picocli;

import io.quarkus.picocli.runtime.PicocliCommandLineFactory;
import io.quarkus.picocli.runtime.annotations.TopCommand;
import picocli.CommandLine;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;

@TopCommand
@CommandLine.Command
public class EntryCommand implements Runnable {
    @CommandLine.Spec
    CommandLine.Model.CommandSpec spec;

    @Override
    public void run() {
        System.out.println("My name is: " + spec.name());
    }
}

@ApplicationScoped
class CustomConfiguration {

    @Produces
    CommandLine customCommandLine(PicocliCommandLineFactory factory) { // <1>
        return factory.create().setCommandName("CustomizedName");
    }
}
----
<1> `PicocliCommandLineFactory` will create an instance of CommandLine with `TopCommand` and `CommandLine.IFactory` injected.

> == Native mode support

Reviewer: Can't we generate this configuration as part of the Quarkus build? I reckon there might be some redundancy but that's the Quarkus way of doing things.

Author: Do you mean that we should have property like `quarkus.picocli.command-name` which would configure this? I can expose most of setters from CommandLine as properties, but there are many, optional, things which can be configured there only programmatically. So I would like to leave `PicocliCommandLineFactory` (maybe with different name) to allow easy reconfiguration of CommandLine. 

I'm also not sure which of CommandLine properties should be configured in runtime. This maybe not a real issue, but for example setting `QUARKUS_PICOCLI_TRIM_QUOTES` in env will change behavior of your command line application, e.g. it will start to reacting to arguments in different way.

Reviewer: No my comment was about the native mode. We shouldn't need an annotation processor to configure the native mode, everything should be done in the Quarkus processor via build items. We can for sure help with this!

Author: I will try to port picocli-codegen code to build steps. Reflection and resource bundles should be straightforward. Not sure about dynamic proxies, but I think I saw code for this also in quarkus.

> Picocli tool provides annotation processor which will generate a configuration for GraalVM Native images.

Reviewer: ```suggestion
The Picocli tool provides an annotation processor which will generate a configuration for GraalVM Native images.
```

> To use this annotation processor you need to add following dependency to `pom.xml` file:

Reviewer: ```suggestion
To use this annotation processor you need to add the following dependency to the `pom.xml` file:
```

[source,xml]

> --

Reviewer: ```suggestion
----
```

<dependency>
  <groupId>info.picocli</groupId>
  <artifactId>picocli-codegen</artifactId>
</dependency>

> --

Reviewer: ```suggestion
----
```

> For gradle you need to add following in `dependencies` section of `build.gradle` file:

Reviewer: ```suggestion
For Gradle you need to add the following in the `dependencies` section of the `build.gradle` file:
```

include::./status-include.adoc[]

IMPORTANT: If you are not familiar with the Quarkus Command Mode, consider reading the link:command-mode-reference[Command Mode reference guide] first.

== Configuration

Once you have your Quarkus project configured you can add the `picocli` extension
to your project by running the following command in your project base directory.

[source,bash]
----
./mvnw quarkus:add-extension -Dextensions="picocli"
----

This will add the following to your pom.xml:

[source,xml]
----
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-picocli</artifactId>
    </dependency>
----

== Simple command line application

Simple PicocliApplication with only one `Command` can be created as follows:

[source,java]
----
package com.acme.picocli;

import picocli.CommandLine;

import javax.enterprise.context.Dependent;
import javax.inject.Inject;

@CommandLine.Command // <1>
public class HelloCommand implements Runnable {

    @CommandLine.Option(names = {"-n", "--name"}, description = "Who will we greet?", defaultValue = "World")
    String name;

    private final GreetingService greetingService;

>     @Inject // <2>

Reviewer 6: Let's drop this annotation as it isn't needed when there is a single constuctor

    @Override
    public void run() {
        System.out.println("Goodbye World!");
    }
}
----

== Customizing Picocli CommandLine instance

You can customize CommandLine classes used by the `picocli` extension by producing your own bean instance:

[source,java]
----
package com.acme.picocli;

import io.quarkus.picocli.runtime.PicocliCommandLineFactory;
import io.quarkus.picocli.runtime.annotations.TopCommand;
import picocli.CommandLine;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;

@TopCommand
@CommandLine.Command
public class EntryCommand implements Runnable {
    @CommandLine.Spec
    CommandLine.Model.CommandSpec spec;

    @Override
    public void run() {
        System.out.println("My name is: " + spec.name());
    }
}

@ApplicationScoped
class CustomConfiguration {

    @Produces
    CommandLine customCommandLine(PicocliCommandLineFactory factory) { // <1>
        return factory.create().setCommandName("CustomizedName");
    }
}
----
<1> `PicocliCommandLineFactory` will create an instance of CommandLine with `TopCommand` and `CommandLine.IFactory` injected.

== Different entry command for each profile

> It is possible to create different entry command for each profile, using `@IfBuildProfile`:

Reviewer 6: Very good idea to mention this!

    @IfBuildProfile("prod")
    public Object prodCommand() {
        return new ProdCommand("Configured by me!");
    }

}
----
<1> You can return instance of `java.lang.Class` here. In such case `CommandLine` will try to instantiate this class using `CommandLine.IFactory`.

== Providing own QuarkusMain

You can also provide your own application entry point annotated with `QuarkusMain` (as described in link:command-mode-reference[Command Mode reference guide]).

[source,java]
----
package com.acme.picocli;

import io.quarkus.runtime.QuarkusApplication;
import io.quarkus.runtime.annotations.QuarkusMain;
import picocli.CommandLine;

import javax.inject.Inject;

@QuarkusMain
@CommandLine.Command(name = "demo", mixinStandardHelpOptions = true)
public class ExampleApp implements Runnable, QuarkusApplication {
    @Inject
    CommandLine.IFactory factory; // <1>

    @Override
    public void run() {
        // business logic
    }

    @Override
    public int run(String... args) throws Exception {
        return new CommandLine(this, factory).execute(args);
    }
}
----
<1> Quarkus-compatible `CommandLine.IFactory` bean created by `picocli` extension.

== Native mode support

> GraalVM Native images should work out-of-box, but in case of any issues it is still possible to use annotation processor from `picocli` project.

Reviewer 6: I am a little hesitant about mentioning this TBH... Nowhere in any other extension are we telling users to add some additional process to get their GraalVM native image working. 
We should strive to make everything work in native image and treat things that don't as bugs to be fixed here, instead of relying on an annotation processor.

Author: I agree that if something doesn't work in native images created with buildsteps then it is a bug and must be fixed. But, if I understood @remkop correctly, `picocli-codegen` may have some improvements in future to make native images smaller by tweaking reflection configuration. With Quarkus, users won't be able to benefit from those improvements until they will be ported to buildsteps, which is more like feature and not bugfix. 

As for now we probably can remove this property and decided what to do when improvements will be added to `picocli-codegen`.

Reviewer 6: Well, in my view if we want to tweak the native image stuff for picocli, then the best way to do that would be to have the proper hooks in picocli so that Quarkus could do it's own optimizations (much in the same way that is already the case for other extensions)

Author: I'm little lost here. Do you mean hooks in `picocli` extension or `picocli` library itself? Current implementation of buildsteps is quite aggressive in adding reflection to native image. For example all methods are added from @Command annotated classes, even if they are not called by reflection in runtime. And @remkop also mentioned possible changes in picocli library itself.

Reviewer 6: I mean that picocli could (might already ?) Have hooks that the Quarkus extension could use to optimize things (create bytecode for implementing interfaces for example).
This is of course something for future consideration

Reviewer 3: > As for now we probably can remove this property and decided what to do when improvements will be added to picocli-codegen.

😭  I had concerns with build steps replacing `picocli-codegen` because this potentially limits how picocli can evolve as a library (without breaking picocli in Quarkus). My concerns were addressed by adding this property. 

The purpose of this property was to provide a fallback solution for any interim periods of time when picocli has evolved in an incompatible way, and the Quarkus picocli extension has not been updated yet. 

The existence of this fallback mechanism is important to me.  Please reinstate it.

> I am a little hesitant about mentioning this TBH... 

Now, how this fallback mechanism is _documented_ is a different matter.
The intention is certainly not to advertise this as canonical or normal usage.

I am fine with very minimal documentation or even omitting this from the documentation.
